### ISSUE TITLE
False positive - null reference

### ISSUE THREAD

**Issue by marcosdsanchez** [kalessil/phpinspectionsea#358]
Reference:

```php
class NullReferencer {
    public function nullReference(): ?self { return new $this; }
}

$lastDetail = new NullReferencer;
```

Given this code ([original image](http://i.imgur.com/doIanOJ.png)):

```php
if ($lastDetail !== null && $lastDetail->nullReference() !== null) {
    $lastDetail->nullReference()->nullReference();
}
```

I wouldn't expect a null reference check in the 2nd line.

This does not happen if I use a temp variable ([original image](http://i.imgur.com/ECNHXVZ.png)):

```php
if ($lastDetail !== null && $lastDetail->nullReference() !== null) {
    $tempVariable = $lastDetail->nullReference();
    $tempVariable->nullReference();
}
```

**Comment by kalessil** [comment]
I want to give the inspection a little bit more intelligence, but meanwhile following should fix the issue 👍 

```php
   $card = null === $lastDetail ? null : $lastDetail->getCard();
   if (null !== card) {
       /* ... */
   }
```

**Comment by rentalhost** [comment]
@kalessil Take a look on [this implementation](https://github.com/rentalhost/laravel-insight/blob/master/src/utils/PhpClassUtil.java#L212). I guess that it could solve problems like that. You should pass any kind of element, and it will try resolve the related types.

**Comment by kalessil** [comment]
Here a little bit another issue, the call is nullable and checked in condition for nullability. Type resolving works properly here.

**Comment by rentalhost** [comment]
Hmm... Now I understand. In this case, it should accept `$a->b()->c()` when `$a->b()` is previously tested and pass. But in all way, I don't know if it is a solution. When you call `$a->b()` for the first time, you can receives a result different of the second time. For instance, a implementation for `$array->pop()`. I think that should be better you uses a variable to store last value (it too should avoid double-call on a function).

**Comment by kalessil** [comment]
It would be the best way, but 100% of feedback so far is about supporting the case.

**Comment by kalessil** [comment]
Ready (fresh binary pushed as well)
